src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/LanguageServer/Handler/Restore/RestoreHandler.cs

```diff
@@ -0,0 +1,107 @@
+    {
+        Contract.ThrowIfNull(context.Solution);
+        using var progress = BufferedProgress.Create(request.PartialResultToken);
+
+        progress.Report(new RestorePartialResult(LanguageServerResources.Restore, LanguageServerResources.Restore_started));
+
+        var restorePaths = GetRestorePaths(request, context.Solution, context);
+        if (restorePaths.IsEmpty)
+        {
+            progress.Report(new RestorePartialResult(LanguageServerResources.Restore, LanguageServerResources.Nothing_found_to_restore));
+            return progress.GetValues() ?? [];
+        }
+
+        await RestoreAsync(restorePaths, progress, cancellationToken);
+
+        progress.Report(new RestorePartialResult(LanguageServerResources.Restore, $"{LanguageServerResources.Restore_complete}{Environment.NewLine}"));
+        return progress.GetValues() ?? [];
+    }
+
+    private async Task RestoreAsync(ImmutableArray<string> pathsToRestore, BufferedProgress<RestorePartialResult> progress, CancellationToken cancellationToken)
+    {
+        foreach (var path in pathsToRestore)
+        {
+            var arguments = $"restore \"{path}\"";
+            var workingDirectory = Path.GetDirectoryName(path);
+            var stageName = string.Format(LanguageServerResources.Restoring_0, Path.GetFileName(path));
+            ReportProgress(progress, stageName, string.Format(LanguageServerResources.Running_dotnet_restore_on_0, path));
+
+            var process = dotnetCliHelper.Run(arguments, workingDirectory, shouldLocalizeOutput: true);
+
+            process.OutputDataReceived += (sender, args) => ReportProgress(progress, stageName, args.Data);
+            process.ErrorDataReceived += (sender, args) => ReportProgress(progress, stageName, args.Data);
+            process.BeginOutputReadLine();
+            process.BeginErrorReadLine();
+
+            await process.WaitForExitAsync(cancellationToken);
+
+            if (process.ExitCode != 0)
+            {
+                throw new InvalidOperationException(LanguageServerResources.Failed_to_run_restore_see_output_for_details);
+            }
+        }
+
+        static void ReportProgress(BufferedProgress<RestorePartialResult> progress, string stage, string? restoreOutput)
+        {
+            if (restoreOutput != null)
+            {
+                progress.Report(new RestorePartialResult(stage, restoreOutput));
+            }
+        }
+    }
+
+    private static ImmutableArray<string> GetRestorePaths(RestoreParams request, Solution solution, RequestContext context)
+    {
+        if (request.ProjectFilePath != null)
+        {
+            return ImmutableArray.Create(request.ProjectFilePath);
+        }
+
+        // No file paths were specified - this means we should restore all projects in the solution.
+        // If there is a valid solution path, use that as the restore path.
+        if (solution.FilePath != null)
```

Comment:
This doesn't need a change now, but we should still think about what happens if the user has a solution file and also has load on demand enabled, since this wouldn't cover the load-on-demand projects.